### ISSUE TITLE
Prepoznavanje Žanra Muzike

### ISSUE THREAD

**Issue by VukoticMarko** [ftn-ai-lab/nm-2024#9]
### Tim

Marko Vukotić R2-45-2023

---

### Definicija problema

Cilj projekta je klasifikacija muzičkih žanrova iz muzičkih mp3 fajlova. Ideja je da se audio signal pre-processuje u vizuelnu reprezentaciju (mel-spektrogram 224×224) i zatim da se trenira neuronska mreža koja na osnovu slike predviđa žanr tog mp3 fajla. 

Projekat ne spajam projektom iz drugog predmeta.

### Skup podataka

Skup podataka je  _FMA (Free Music Archive) — large_  (~100k MP3 fajlova, ~100GB). _tracks.csv_ se koristi kao metadata za FMA database i koristi se za kreiranje podataka koji će se ubacivati u mrežu.  Učitavanje bi se vršilo sa skriptom gde se svaki MP3 fajl sa librosa bibliotekom učitava. Za svaki segment izračunava se mel-spektrogram sa parametrima pa se zatim normalizuje i skalira u opseg [0,255] i na kraju konvertuje u 8-bitnu sliku. Slika će biti u 224x224 px (Res-Net Veličina). Slike će najverovatnije biti grayscale (224,224,1).

Podešavanje _data processing_ skripte:
```
AUDIO_ROOT = "/media/fma_large/fma_large"
TRACKS_CSV = "datasets/fma_metadata/tracks.csv"
SR = 22050
DURATION = 30  # Seconds
NUM_SEGMENTS = 4
# STFT settings
N_MELS = 128
N_FFT = 2048
HOP_LENGTH = 512
IMG_SIZE = (224, 224)
```

### Metodologija

Preprocesiranje: učitavanje MP3 → normalizacija/padding → segmentacija trake i odabir dužine uzorka u sekundama → mel-spectogram → filtriranje tiših/loših segmenata → export u slike dimenziji 224×224.

Augmentacija: RandomFlip, RandomZoom, RandomContrast, RandomBrightness iz kerasove biblioteke.

Model: CNN (sekvencijalni model sa više Conv2D slojeva, BatchNorm, Dropout, GlobalAveragePooling) sa L2 regularizacijom. 
Izlaz je kategorijalna softmax klasifikacija sa num_classes koji se dinamički određuje iz folder strukture dataseta (ovom metodom radiće svi FMA datasetovi).

Trening će se vršiti na lokalnoj mašini sa GPU (Linux sa Cuda i cuDNN) koristeći mixed precision (mixed_float16). Model će koristiti Adam optimizator. Kao mehanizme za kontrolu treninga koristiću dva callback-a: EarlyStopping (prati val_loss, patience~20,) koji vraća najbolje težine i ReduceLROnPlateau (smanjuje stopu učenja za faktor ako val_loss ne opadne u nekom patience). Trening se pokreće pozivom model.fit sa generatorima za trening i validaciju i navedenim callback-ima. Hiperparametri će biti definisani sa:

```
BATCH_SIZE = 4
INPUT_SHAPE = (224, 224)
CNN_INPUT = (224, 224, 1) # 3rd parameter: 1 - gray | 3 - RGB
EPOCHS = 400
MODEL_NAME = "model-projekat-neuronske-mreze-master"
REGULARLIZER = tf.keras.regularizers.L2(1e-4)
```

### Evaluacija

Dataset se deli na trening/validaciju (validation_split=0.15 → 85%/15%). Primarna mera performansi je **accuracy**.

**Comment by vdragan1993** [comment]
1. Nije poznata dimenzionalnost problema, odnosno koliko žanrova će se klasifikovati.
2. Pored navedenog modela, potrebno je ubaciti još jedan model sa kojim će se porediti rezultat. Predlažem da se za to koristi Transfer Learning.
3. `accuracy` se ne može koristiti kao mera performansi. Potrebno je koristiti prikladnije metrike i na nivou celokupnog problema, a i na nivou klasa.
4. `Potencijalni Problemi` se ne traže u prijavi projekta i treba ih ukloniti.

**Comment by VukoticMarko** [comment]
1. Za sad je plan 15 dimenzija. Takođe plan je da ne bude hardcodeovan broj dimenzija nego da može biti dinamički na osnovu dataseta mp3 fajlova koji se učitava. Gledaću metadata **genre_top** u tracks.csv u kojima su grupisane sve pesme (kao neki high-level žanrovi. Primer: postoji _electronic_ žanr u kojem će biti svi žanrovi elektronske muzike kao EDM, Techno itd.) Bez gledanja genre_top se dobije ogroman broj dimenzija koje su jako nebalansirane.
2. Da li je okej da ga poredim sa ResNet (pošto sam već optimizovao veličinu slike da bude resnet compatible)?
3. Za celokupni problem da li je okej da dodam i koristim neke od Macro F1, Weighted F1, balanced accuracy ili top-k accuracy?
Za klasnu da dodam Precision Recall F1? Confusion Matrix?
4. Uklonjeno

**Comment by vdragan1993** [comment]
1. Da bi projekat bio precizno definisan potrebno je znati tačan broj dimenzija/klasa sa kojima će se raditi.
2. Jeste.
3. Jeste.

**Comment by VukoticMarko** [comment]
1. 15 klasa
`blues
classical
country
easy listening
electronic
experimental
folk
hip-hop
instrumental
international
jazz
pop
rock
soul-rnb
spoken`

**Comment by vdragan1993** [comment]
Tema odobrena.
Ostavite link ka GitHub repozitorijumu projekta.
Srećan rad.

**Comment by VukoticMarko** [comment]
https://github.com/VukoticMarko/music-genre-classification
